File: ios/CloudPaymentsApi/CloudPaymentsApi.m
//
//  CloudPaymentsApi.m
//  CloudpaymentsSdk
//
//  Created by Anton Votinov on 05.02.2022.
//  Copyright © 2022 Facebook. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(CloudPaymentsApi, NSObject)
  RCT_EXTERN_METHOD(initialization: (NSDictionary*)paymentData)
  RCT_EXTERN_METHOD(auth: (NSString*)cardCryptogramPacket email: (NSString*)email resolve: (RCTPromiseResolveBlock)resolve reject: (RCTPromiseRejectBlock)reject)
  RCT_EXTERN_METHOD(charge: (NSString*)cardCryptogramPacket email: (NSString*)email resolve: (RCTPromiseResolveBlock)resolve reject: (RCTPromiseRejectBlock)reject)
  RCT_EXTERN_METHOD(setInformationAboutPaymentOfProduct: (NSDictionary*)details)
@end

File: ios/CloudPaymentsAPI/CloudPaymentsAPI.swift
import Foundation
import React
import Cloudpayments;

@objc(CloudPaymentsAPISwift)
public class CloudPaymentsAPISwift: NSObject {
  private var api: CloudpaymentsApi?;
  
  @objc
  public static let shared = CloudPaymentsAPISwift()
  
  @objc
  private override init() {
    super.init()
  }
  
  @objc
  public func initialization(_ publicId: String) {
    self.api = CloudpaymentsApi.init(publicId: publicId)
  }
  
  @objc
  public func getPublicKey(_ resolve: RCTPromiseResolveBlock?, reject: RCTPromiseRejectBlock?) {
    reject?("error", "method not yet available", nil);
  }
  
  @objc
  public func getBinInfo(_ cardNumber: String, resolve: RCTPromiseResolveBlock?, reject: RCTPromiseRejectBlock?) {
    CloudpaymentsApi.getBankInfo(cardNumber: cardNumber) { (info, error) in
        if let error = error {
          reject?("error", error.message, nil);
        } else {
          var bankInfo: Dictionary<String, Any?> = [:];
          
          let infoConverted = info as! BankInfo?
          
          bankInfo["cardType"] = infoConverted?.cardType
          bankInfo["bankName"] = nil
          bankInfo["logoUrl"] = infoConverted?.logoURL
          bankInfo["currency"] = infoConverted?.currency
          bankInfo["convertedAmount"] = infoConverted?.convertedAmount
          bankInfo["hideCvv"] = infoConverted?.hideCvvInput
          
          resolve?(bankInfo);
        }
    }
  }
}
